src/black/__init__.py

```diff
@@ -5767,44 +5801,34 @@ def gen_python_files(
     report: "Report",
     gitignore: PathSpec,
 ) -> Iterator[Path]:
-    """Generate all files under `path` whose paths are not excluded by the
-    `exclude` regex, but are included by the `include` regex.
+    """Generate all files under `path` whose paths are not excluded by
+    `exclude_regexes`, but are included by the `include` regex.
 
     Symbolic links pointing outside of the `root` directory are ignored.
 
     `report` is where output about exclusions goes.
     """
     assert root.is_absolute(), f"INTERNAL ERROR: `root` must be absolute but is {root}"
     for child in paths:
-        # Then ignore with `exclude` option.
-        try:
-            normalized_path = child.resolve().relative_to(root).as_posix()
-        except OSError as e:
-            report.path_ignored(child, f"cannot be read because {e}")
+        normalized_path = normalize_path_maybe_ignore(child, root, report)
+        if normalized_path is None:
             continue
-        except ValueError:
-            if child.is_symlink():
-                report.path_ignored(
-                    child, f"is a symbolic link that points outside {root}"
-                )
-                continue
-
-            raise
 
         # First ignore files matching .gitignore
         if gitignore.match_file(normalized_path):
             report.path_ignored(child, "matches the .gitignore file content")
             continue
 
+        # Then ignore with `exclude` and `--force-exclude` options.
         normalized_path = "/" + normalized_path
         if child.is_dir():
             normalized_path += "/"
 
         is_excluded = False
-        for exclude in exclude_regexes:
+        for opt_name, exclude in zip(("--exclude", "--force-exclude"), exclude_regexes):
```

Comment:
This makes me a bit uneasy because it assumes that `exclude_regexes` exactly matches these two flags, when the type doesn't demand that. It would be cleaner to make exclude_regex and force_exclude_regex separate arguments to this function, or to make `exclude_regexes` an iterable of (option name, pattern) pairs.

Reply:
Yeah, this is not my cleanest code ever. I have already thought about moving the force exclude regex into its own argument but I decided not doing so as it would have make more sense when `--extend-exclude` is implemented. No problems doing it now than later!